File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def help_function(self, root:Optional[TreeNode], small: int, big: int):
        if root == None:
            return True
        if small >= root.val or big <= root.val:
            return False
        return self.help_function(root.left, small, root.val) and self.help_function(root.right, root.val, big)

    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.help_function(root, -1001, 1001)

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.empty() || inorder.empty())return nullptr;
        map<int, int> mp;
        for(auto i=0; i < inorder.size(); i++)
        {
            mp[inorder[i]] =i;
        }

        TreeNode* root = new TreeNode(preorder[0]);
        int mid = mp[preorder[0]];
        //left subtree
        vector<int>leftP(preorder.begin()+1, preorder.begin()+1+mid);
        vector<int>leftI(inorder.begin(), inorder.begin()+mid);
        root->left = buildTree(leftP, leftI);
        //right subtree
        vector<int>rightP(preorder.begin()+1+mid, preorder.end());
        vector<int>rightI(inorder.begin()+mid+1, inorder.end());
        root->right = buildTree(rightP, rightI);

        return root;
    }

    
};